aws/resource_aws_iot_topic_rule.go

```diff
@@ -11,419 +11,433 @@ import (
 )
 
 func resourceAwsIotTopicRule() *schema.Resource {
-	return &schema.Resource{
-		Create: resourceAwsIotTopicRuleCreate,
-		Read:   resourceAwsIotTopicRuleRead,
-		Update: resourceAwsIotTopicRuleUpdate,
-		Delete: resourceAwsIotTopicRuleDelete,
-
-		Importer: &schema.ResourceImporter{
-			State: schema.ImportStatePassthrough,
-		},
-
-		Schema: map[string]*schema.Schema{
-			"arn": {
-				Type:     schema.TypeString,
-				Computed: true,
-			},
-			"cloudwatch_alarm": {
-				Type:     schema.TypeSet,
-				Optional: true,
-				Elem: &schema.Resource{
-					Schema: map[string]*schema.Schema{
-						"alarm_name": {
-							Type:     schema.TypeString,
-							Required: true,
-						},
-						"role_arn": {
-							Type:         schema.TypeString,
-							Required:     true,
-							ValidateFunc: validateArn,
-						},
-						"state_reason": {
-							Type:     schema.TypeString,
-							Required: true,
-						},
-						"state_value": {
-							Type:         schema.TypeString,
-							Required:     true,
-							ValidateFunc: validateIoTTopicRuleCloudWatchAlarmStateValue,
-						},
+	actionsSchema := map[string]*schema.Schema{
```

Comment:
Since we will very likely want to deprecate the root attributes and since we typically prefer static schema declarations, it would be great if the existing schema was left alone and the new `error_action` handling was additive (copy-paste is fine) for now. 👍

Reply:
@bflad Thanks for the review. One thing I did not know how to do for the `error_actions` block was to limit it to only 1 action, since the AWS API accepts multiple actions and only a single error action. Can you point me in the direction to add that validation to the schema?

Reply:
Absolutely! You can use a combination of a top level attribute of `Type: schema.TypeList` + `MaxItems: 1` then each of the nested attributes can set [`ExactlyOneOf`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/schema?tab=doc#Schema) (this will require re-defining the same attribute list for each attribute since this validation functionality is per-attribute, but we have plans to move this to the resource level at some point), e.g.

```go
// Quick conceptual schema implementation
"error_action": {
  Type: schema.TypeList,
  Optional: true,
  MaxItems: 1,
  Elem: &schema.Resource{
    Schema: map[string]*schema.Schema{
      "cloudwatch_alarm": {
         Type: schema.TypeList,
         Optional: true,
         MaxItems: 1,
         Elem: &schema.Resource{ /* ... */ },
         ExactlyOneOf: []string{
           "error_action.0.cloudwatch_alarm",
           "error_action.0.dynamodb",
           "error_action.0.dynamodbv2",
           /* ... */
         },
      },
      "dynamodb": {
         Type: schema.TypeList,
         Optional: true,
         MaxItems: 1,
         Elem: &schema.Resource{ /* ... */ },
         ExactlyOneOf: []string{
           "error_action.0.cloudwatch_alarm",
           "error_action.0.dynamodb",
           "error_action.0.dynamodbv2",
           /* ... */
         },
      },
    },
  },
}
```

As shown above for the new schema it would be nice to have the underlying configuration blocks using `TypeList` and `MaxItems: 1` as well to prevent multiple configuration blocks. Just note that the attribute type assertions via `.(*schema.Set).List()` will become just `.([]interface{})`. 😄